Implement Generate Hierarchical CV DB

itsfarseen opened the pull request:
Depends on #13

itsfarseen committed:
cv_hierarchical_storage: fix bug in verifying configuration

itsfarseen committed:
cv_hierarchical_storage: Add message to help debug assertion failure

itsfarseen committed:
generate_hierarchical_cv_db: implement

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+# 
+
+def main():
+    compressor_chkpt = "../rigel-data/hierarchical-compression-checkpoint-2024-10-30/attention_model.pt"
```
This should be an argument passed to `main`

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+    compressor = Compressor(compressor_chkpt)
+    #batch_size = 1024
+
+    input_db_path = Path("../rigel-data/context-vectors-2024-10-30/")
+    output_db_path = Path("../rigel-data/context-vectors-compressed-2024-10-30")
```
These should also be arguments

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+from cv_library.compressor import Compressor
```
I'd actuall prefer this not to be a module, but a runnable file `generate_hierarchical_cv_db.py` in the root

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+    return res
+
+
+# Not used because need to figure out how to batch tensors of different sizes
+# def make_batches(xs, batch_size):
+#     for i in range(0, len(xs), batch_size):
+#         yield xs[i:i+batch_size]
+# 
```
```suggestion
```

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+    input_db = ContextVectorDB(input_db_path)
+    output_db_config = cvhs.DBConfig(
+            vec_max_size = 4096,
+            vec_min_size = 8,
```
This value should be, by default, `cv_library.hierarchical_compression.Attention.NUM_HEADS`

itsfarseen replied:
NUM_HEADS seem to be 32. I'm not sure how to use it here.

I've also changed the parameters to DBConfig. Now there's a single `level_sizes` parameter, which is specified in the `justfile` as:

```
--level_sizes 16,64,256,1024,4096
```

Does this work?

itsfarseen replied:
I got these values by inspecting the shape of the compressor's output.

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+    output_db_config = cvhs.DBConfig(
+            vec_max_size = 4096,
+            vec_min_size = 8,
+            compression_factor = 8,
```
This should also be a parameter

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+            vec_min_size = 8,
+            compression_factor = 8,
+            compression_dimension = 1, # cv.shape = (seq_size, 4096). We compress the one with 4096.
+            search_narrow_factor = 16,
```
This should also be a parameter

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+
+    print("Reading metadata ..")
+    headings = get_all_headings(input_db)
+    # headings_batched = list(make_batches(headings, batch_size))
```
```suggestion
```

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+    # headings_batched = list(make_batches(headings, batch_size))
+    
+    print("Processing ..")
+    # for batch in tqdm(headings_batched):
+    #     cvs = []
+    #     for article_title, section_name in batch:
+    #         cvs.append(input_db.get(article_title, section_name))
+    #     cvs = np.stack(cvs)
+    #     cvs = torch.tensor(cvs)
+    #     print("cvs", cvs.shape)
+    #     output_cvs = compressor.compress(cvs)
+    #     print("output_cvs", output_cvs.shape)
```
```suggestion
```

nalane commented on generate_hierarchical_cv_db/__main__.py:
```diff
@@ -0,0 +1,81 @@
+
+
+if __name__ == "__main__":
+    main()
```
You can use Fire to make it so that the parameters to this main function become command line arguments

itsfarseen committed:
cv_storage: extract some common bits to indexed_binary_db to be used in cv_hierarchical_storage

itsfarseen committed:
bin_storage: remove

Functionality provided by indexed_binary_db

itsfarseen committed:
cv_hier_storage: Fix bugs, rename

Rename cv_hierarchical_storage -> cv_hier_storage
Fix bugs and tests

itsfarseen committed:
cv_storage: extract models.py

itsfarseen committed:
indexed_binary_db: Better function names

itsfarseen committed:
generate_hier_cv_db: Rename, fix bugs

Move generate_hierarchical_cv_db/__main__.py to generate_hier_cv_db.py
Fix bugs

itsfarseen committed:
justfile: Add cv_hier_db_xx commands

cv_hier_db_gen
cv_hier_db_verify
cv_hier_db_clean
cv_hier_db_test_e2e

itsfarseen committed:
README: update

nalane commented:
Looks great!
